weasel/_util/general.py

```diff
@@ -0,0 +1,568 @@
+import hashlib
+import logging
+import os
+import shlex
+import shutil
+import subprocess
+import sys
+import tempfile
+import warnings
+from contextlib import contextmanager
+from pathlib import Path
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Dict,
+    Generator,
+    Iterable,
+    Iterator,
+    List,
+    Mapping,
+    NoReturn,
+    Optional,
+    Pattern,
+    Set,
+    Tuple,
+    Type,
+    Union,
+    cast,
+)
+
+import srsly
+import typer
+from confection import Config
+from click import NoSuchOption
+from click.parser import split_arg_string
+from packaging.version import Version
+from cloudpathlib import CloudPath
+from pydantic import BaseModel
+from wasabi import msg
+
+from weasel import __version__ as weasel_version
+
+from ..info import COMMAND, PROJECT_FILE, PROJECT_LOCK
+from .compat import is_windows
+from .errors import Warnings
+from .schema import (
+    ProjectConfigSchema,
+    show_validation_error,
+    validate,
+    validate_project_commands,
+    validate_project_version,
+)
+
+Arg = typer.Argument
+Opt = typer.Option
+
+
+logger = logging.getLogger("weasel")
+logger_stream_handler = logging.StreamHandler()
+logger_stream_handler.setFormatter(
+    logging.Formatter("[%(asctime)s] [%(levelname)s] %(message)s")
+)
+logger.addHandler(logger_stream_handler)
+
+
+class SimpleFrozenDict(dict):
```

Comment:
This seems to be essentially the same as `SimpleFrozenList` below: do we really want both?

Reply:
My original suggestion was to include neither of these in `weasel` and use `None` + internal empty list/dict assignment.

Reply:
I removed both the SimpleFrozen things.